source/virtualservice.go

```diff
@@ -0,0 +1,445 @@
+	err = wait.Poll(time.Second, 60*time.Second, func() (bool, error) {
+		return serviceInformer.Informer().HasSynced(), nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to sync cache: %v", err)
+	}
+
+	return &virtualServiceSource{
+		kubeClient:               kubeClient,
+		istioClient:              istioClient,
+		namespace:                namespace,
+		annotationFilter:         annotationFilter,
+		fqdnTemplate:             tmpl,
+		combineFQDNAnnotation:    combineFQDNAnnotation,
+		ignoreHostnameAnnotation: ignoreHostnameAnnotation,
+		serviceInformer:          serviceInformer,
+	}, nil
+}
+
+// Endpoints returns endpoint objects for each host-target combination that should be processed.
+// Retrieves all VirtualService resources in the source's namespace(s).
+func (sc *virtualServiceSource) Endpoints() ([]*endpoint.Endpoint, error) {
+	virtualServiceList, err := sc.istioClient.NetworkingV1alpha3().VirtualServices(sc.namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	virtualServices := virtualServiceList.Items
+	virtualServices, err = sc.filterByAnnotations(virtualServices)
+	if err != nil {
+		return nil, err
+	}
+
+	var endpoints []*endpoint.Endpoint
+
+	for _, virtualService := range virtualServices {
+		// Check controller annotation to see if we are responsible.
+		controller, ok := virtualService.Annotations[controllerAnnotationKey]
+		if ok && controller != controllerAnnotationValue {
+			log.Debugf("Skipping VirtualService %s/%s because controller value does not match, found: %s, required: %s",
+				virtualService.Namespace, virtualService.Name, controller, controllerAnnotationValue)
+			continue
+		}
+
+		gwEndpoints, err := sc.endpointsFromVirtualService(virtualService)
+		if err != nil {
+			return nil, err
+		}
+
+		// apply template if host is missing on VirtualService
+		if (sc.combineFQDNAnnotation || len(gwEndpoints) == 0) && sc.fqdnTemplate != nil {
+			iEndpoints, err := sc.endpointsFromTemplate(virtualService)
+			if err != nil {
+				return nil, err
+			}
+
+			if sc.combineFQDNAnnotation {
+				gwEndpoints = append(gwEndpoints, iEndpoints...)
+			} else {
+				gwEndpoints = iEndpoints
+			}
+		}
+
+		if len(gwEndpoints) == 0 {
+			log.Debugf("No endpoints could be generated from VirtualService %s/%s", virtualService.Namespace, virtualService.Name)
+			continue
+		}
+
+		log.Debugf("Endpoints generated from VirtualService: %s/%s: %v", virtualService.Namespace, virtualService.Name, gwEndpoints)
+		sc.setResourceLabel(virtualService, gwEndpoints)
+		endpoints = append(endpoints, gwEndpoints...)
+	}
+
+	for _, ep := range endpoints {
+		sort.Sort(ep.Targets)
+	}
+
+	return endpoints, nil
+}
+
+func (sc *virtualServiceSource) AddEventHandler(ctx context.Context, handler func()) {
+}
+
+func (sc *virtualServiceSource) getGateway(gatewayStr string, virtualService networkingv1alpha3.VirtualService) *networkingv1alpha3.Gateway {
+	if gatewayStr == "" || gatewayStr == "mesh" {
```

Comment:
This also seems like its dev's internal convention leaking in the code. Since gateway str could be name differently. Ideally we should through an error if its empty or non existent.

Reply:
@rverma-jm that is a good point. Ideally, this logic would need to be shipped by Istio itself through a library or through some filtering option in `client-go`. Might be worth opening a separate issue in `istio/istio` for this.

Reply:
@Raffo sure